Assets/Scripts/Ports/PortCommunicationManager.cs

```diff
@@ -0,0 +1,258 @@
+            var tcpEndPoint = new IPEndPoint(IPAddress.Any, port);
+            tcpSocket.Bind(tcpEndPoint);
+            tcpSocket.Listen(1);
+            currentSock = tcpSocket;
+
+            try
+            {
+                using var client = tcpSocket.Accept();
+                CommunicationLoop(client);
+            }
+            catch
+            {
+                // Thread stopped before connection
+            }
+        }
+
+        private static void UDPThread(int port)
+        {
+            using var udpSocket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            var udpEndPoint = new IPEndPoint(IPAddress.Any, port);
+            udpSocket.Bind(udpEndPoint);
+            currentSock = udpSocket;
+
+            byte[] tempBuffer = new byte[BUFFER_SIZE];
+            EndPoint remoteEP = new IPEndPoint(IPAddress.Any, 0);
+            bool connected = false;
+
+            while (running)
+            {
+                try 
+                {
+                    // UDP is connectionless...
+                    if (udpSocket.Poll(0, SelectMode.SelectRead))
+                    {
+                        int received = udpSocket.ReceiveFrom(tempBuffer, ref remoteEP);
+                        if (received > 0)
+                        {
+                            Buffer.BlockCopy(tempBuffer, 0, inputBuffers[1 - activeInputBuffer], 0, BUFFER_SIZE);
+                            activeInputBuffer = 1 - activeInputBuffer;
+                            connected = true;
+                        }
+                    }
+
+                    // ... but we still need a connection!
+                    if (connected)
+                    {
+                        udpSocket.SendTo(outputBuffer, remoteEP);
+                    }
+                }
+                catch
+                {
+                    
+                }
+            }
+        }
+
+        private static void CommunicationLoop(Socket client)
+        {
+            byte[] tempBuffer = new byte[BUFFER_SIZE];
+            while (running)
+            {
+                // 1. Read external input
+                if (client.Poll(0, SelectMode.SelectRead))
+                {
+                    if (client.Receive(tempBuffer) > 0)
+                    {
+                        Buffer.BlockCopy(tempBuffer, 0, inputBuffers[1 - activeInputBuffer], 0, BUFFER_SIZE);
+                        activeInputBuffer = 1 - activeInputBuffer;
+                    }
+                }
+
+                // 2. Write output
+                client.Send(outputBuffer);
+            }
+            client.Close();
+        }
+
+        public static void Shutdown()
+        {
+#if UNITY_EDITOR_WIN || UNITY_STANDALONE_WIN
+            winPipe?.Close();
+#endif
+            currentSock?.Close();
+            running = false;
+            commThread?.Join(50);
```

Comment:
This is somewhat prone to so-called UAF (Use-After-Free) issues, as you can't be sure the resource is still being used at that time. This absolutely bags for hard-to-find race conditions. Instead ensure we are the only ones accessing this resource first, before releasing it to the OS:

```suggestion
            running = false;
            commThread?.Join(50);
#if UNITY_EDITOR_WIN || UNITY_STANDALONE_WIN
            winPipe?.Close();
#endif
            currentSock?.Close();
```

This implies that each thread only exits after it completed all of its tasks on the socket, which seems to be the case from the implementation above. Note however, that with TCP listening sockets, there's actually two resources: The listener and the actual client connection. We only handle the listener here; the TCP thread is responsible for the client.